### PR TITLE
Fix pagination bug

### LINKED ISSUE #240: Pagination not working as expected with /collections/{identifier}/products
### Checked for duplicates

Yes - I've already checked

### 🐛 Describe the bug

@jordanpadams thought the [deep-archive](https://github.com/NASA-PDS/deep-archive/) was paginating incorrectly. I know it worked at one time.

The [documentation](https://nasa-pds.github.io/pds-api/specifications/search-v1.0.0-redoc.html#operation/products-of-a-collection) says the `limit` is a cardinal, telling _how many_ to return. But he suspected that it might now be an ordinal, telling the _end index_ of the row to return.

I [tested both extensively in a comment on a separate issues](https://github.com/NASA-PDS/deep-archive/issues/140#issuecomment-1404269589) and found that it's neither. It's sort of just broken.

### 🕵️ Expected behavior

I expected to paginate by using `start` and `limit` parameters, preferably with `limit` telling _how many_ to return.

### 📜 To Reproduce

```bash
See https://github.com/NASA-PDS/deep-archive/issues/140#issuecomment-1404269589
```

### 🖥 Environment Info

```bash
- https://pds.nasa.gov/api/search/1.0 as of 2023-01-26
```

### 📚 Version of Software Used

- https://pds.nasa.gov/api/search/1.0 as of 2023-01-26

### 🩺 Test Data / Additional context

_No response_

### 🦄 Related requirements

N/A

### ⚙️ Engineering Details

May be tightly coupled with https://github.com/NASA-PDS/registry-api/issues/231

### PR DESCRIPTION
**Pull request by alexdunnjpl**
## 🗒️ Summary
Fixes a bug in `PaginationLidvidBuilder` which resulted in incorrect slicing, as well as crashes when provided `start` parameter was greater than provided `limit` 

## ⚙️ Test Data and/or Report
Added comprehensive unit tests for `PaginationLidvidBuilder.addAll()`

## ♻️ Related Issues
Fixes #240 

#250 outstanding

### CODE REVIEW

File: service/src/test/java/gov/nasa/pds/api/registry/model/PaginationLidvidBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        for (List<String> input : complexInput) {
+            pageBuilder.addAll(input);
+        }
+        Assert.assertArrayEquals(new String[]{"8"}, pageBuilder.page().toArray());
+    }
+
+
+    @Test
+    public void testComplexAddTrimHeadAndTailSpanning() {
+        PaginationLidvidBuilder pageBuilder = new PaginationLidvidBuilder(new LidvidsContextStub(2, 5));
```
Comment (nutjob4life) [inline service/src/test/java/gov/nasa/pds/api/registry/model/PaginationLidvidBuilderTest.java:236]: Try any negatives like `LidvidsContextStub(2, -5)` or `LidvidsContextStub(-2, -5)`?

**Review by nutjob4life — APPROVED** [review]
👍

File: service/src/test/java/gov/nasa/pds/api/registry/model/PaginationLidvidBuilderTest.java (hunk 1 above)
Comment (alexdunnjpl) [reply service/src/test/java/gov/nasa/pds/api/registry/model/PaginationLidvidBuilderTest.java:236]: Seemed defensible (albeit naive) to assume that a passed `LidvidsContext` would be valid.  If that validation isn't already baked into the existing `LidvidsContext` implementations I'll see what I can do there - _might_ be trivial but I need to verify that start/limit aren't used in any context where negatives are (say) expected, but handled after instantiation.

Plus I'll need to see how to throw in a way that will return a 422, which may be more trouble than is worth doing right this second (but I'll add it to the backlog).

Comment (tloubrieu-jpl) [reply service/src/test/java/gov/nasa/pds/api/registry/model/PaginationLidvidBuilderTest.java:236]: Ok to test raise an exception when limit or start is negative. Ok also to keep 422 error for later. 400 with a dedicated message sounds more than enough to me for now. Thanks

**Comment by alexdunnjpl** [comment]
Merging, and will open a new issue for lower bounds (as it's out of scope for 240) and fix asap.

**Pull request merged**